elastic/apm-server
# Fixes issues with transaction metrics aggregation.

## Motivation/summary

The PR fixes the following issues:
* Resets limits to 0 after each publish of the transction metrics.
* Uses correct interval for calculating timestamp of overflow metrics.
* Fixes the monitoring and logging to correctly report the reason for overflow.

Along with the above fixes, the PR also refactors the transaction metrics aggregator to be more readable and maintainable while keeping the amount of changes minimum for backport. Would be happy to hear other opinions and suggestions.

## Checklist

~~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~~
~~- [ ] Documentation has been updated~~

## How to test these changes

The fix has a unit test. Can also be tested by following the below steps:

1. Create a 1 GB deployment of APM-Server.
2. Generate load such that x transactions are sent per second. Keep x to be lower than 500/60 to ensure that the limit of per service transaction group should not be breached.
3. Assert that the limits are NOT breached even after (500/x)*time.Minute.


## Related issues

Fixes #10349
